src/main/java/io/druid/data/input/impl/JSONFlattenParser.java

```diff
@@ -0,0 +1,223 @@
+        return null;
+      }
+      if (node.isIntegralNumber()) {
+        if (node.canConvertToLong()) {
+          return node.asLong();
+        } else {
+          return node.asDouble();
+        }
+      }
+      if (node.isFloatingPointNumber()) {
+        return node.asDouble();
+      }
+
+      final String s = node.asText();
+      final CharsetEncoder enc = Charsets.UTF_8.newEncoder();
+      if (s != null && !enc.canEncode(s)) {
+        // Some whacky characters are in this string (e.g. \uD900). These are problematic because they are decodeable
+        // by new String(...) but will not encode into the same character. This dance here will replace these
+        // characters with something more sane.
+        return new String(s.getBytes(Charsets.UTF_8), Charsets.UTF_8);
+      } else {
+        return s;
+      }
+    }
+  };
+
+  public JSONFlattenParser(
+      ObjectMapper objectMapper,
+      Iterable<String> fieldNames,
+      String timeColumn,
+      ParseFlattenSpec flattenSpec,
+      Function<JsonNode, Object> valFn
+  )
+  {
+    this(objectMapper, fieldNames, null, timeColumn, flattenSpec, valFn);
+  }
+
+  public JSONFlattenParser(
+      ObjectMapper objectMapper,
+      Iterable<String> fieldNames,
+      Iterable<String> exclude,
+      String timeColumn,
+      ParseFlattenSpec flattenSpec,
+      Function<JsonNode, Object> valFn
+  )
+  {
+    this.objectMapper = objectMapper;
+    if (fieldNames != null) {
+      setFieldNames(fieldNames);
+    }
+    this.fieldNames.add(timeColumn);
+    this.exclude = exclude != null ? Sets.newHashSet(exclude) : Sets.<String>newHashSet();
+    this.valFn = valFn == null ? flattenValueFunction : valFn;
+    this.flattenSpec = flattenSpec;
+    for (String metric : flattenSpec.getMetrics()) {
+      this.fieldNames.add(metric);
+    }
+  }
+
+  @Override
+  public List<String> getFieldNames()
+  {
+    return fieldNames;
+  }
+
+  @Override
+  public void setFieldNames(Iterable<String> fieldNames)
+  {
+    ParserUtils.validateFields(fieldNames);
+    this.fieldNames = Lists.newArrayList(fieldNames);
+  }
+
+  public ObjectMapper getObjectMapper()
+  {
+    return objectMapper;
+  }
+
+  public Function<JsonNode, Object> getValFn()
+  {
+    return valFn;
+  }
+
+  public void setValFn(Function<JsonNode, Object> valFn)
+  {
+    this.valFn = valFn;
+  }
+
+  @Override
+  public Map<String, Object> parse(String input)
+  {
+    try {
+      Map<String, Object> map = new LinkedHashMap<>();
+      JsonNode root = objectMapper.readTree(input);
+
+      Iterator<String> keysIter = (fieldNames == null ? root.fieldNames() : fieldNames.iterator());
+
+      while (keysIter.hasNext()) {
+        String key = keysIter.next();
+
+        if (exclude.contains(key)) {
+          continue;
+        }
+
+        JsonNode node;
+        String nestPrefix = flattenSpec.getNestPrefix();
+        if (nestPrefix != null && key.startsWith(nestPrefix)) {
```

Comment:
This re-parses the flatten specs every time the parser parses an object. It should be faster to convert the flatten specs into some intermediate kind of object in the constructor of the JSONFlattenParser, and then use those in `parse`.